### PR TITLE
Mixtral to NeMo conversion script.

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
- [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -104,11 +105,23 @@
     HAVE_TE = False
 
 
-def get_specs(spec_name):
-    name_spec_dict = {"": get_gpt_layer_with_transformer_engine_spec(), "megatron_falcon_gpt": get_falcon_layer_spec()}
-    if spec_name not in name_spec_dict:
+@cache
+def mcore_supports_moe() -> bool:
+    mcore_version = packaging.version.Version(version("megatron-core"))
```
Comment (akoumpa) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:110]: I'm not 100% percent this is the correct package name, but this https://pypi.org/project/megatron-core/ seemed up-to-date.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1547,7 +1560,7 @@ def build_transformer_config(self) -> TransformerConfig:
             For attributes in TransformerConfig that are not in the nemo model config, we add custom logic.
         """
 
-        normalization = self.cfg.get('normalization', 'layernorm')
+        normalization = self.cfg.get('normalization', 'layernorm').lower()
```
Comment (akoumpa) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1563]: if we merge mistral first I should rebase this line.

**Comment by cdj0311** [comment]
Could you provide the convert NeMo to huggingface script?

**Comment by akoumpa** [comment]
@cdj0311 will try to have the NeMo to HF by next week

File: scripts/nlp_language_modeling/convert_nemo_mixtral_to_hf.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    dummy_trainer = Trainer(devices=1, accelerator='cpu', strategy=NLPDDPStrategy())
+    model_config = MegatronGPTModel.restore_from(in_file, trainer=dummy_trainer, return_config=True)
+    model_config.tensor_model_parallel_size = 1
+    model_config.pipeline_model_parallel_size = 1
+    cpu_only = True
+    if cpu_only:
+        map_location = torch.device('cpu')
+        model_config.use_cpu_initialization = True
+    else:
+        map_location = None
```
[the remark this answers is not in the document]
Comment (ericharper) [reply scripts/nlp_language_modeling/convert_nemo_mixtral_to_hf.py:89]: Why do we have this unreachable statement?

[the remark this answers is not in the document]
Comment (akoumpa) [reply scripts/nlp_language_modeling/convert_nemo_mixtral_to_hf.py:89]: I assumed it's a false positive

**Comment by ericharper** [comment]
jenkins

**Comment by ericharper** [comment]
jenkins

**Comment by ericharper** [comment]
jenkins

**Comment by ericharper** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Comment by janekl** [comment]
Hi @ericharper. I browsed CodeQL suggestions in this pull request. This is a good example when **Flake8** is better than CodeQL: CodeQL fails to detect undefined variables (a simple bug), see `head_num` [here](https://github.com/NVIDIA/NeMo/blob/6865c39d85c4f3b120a02fe74024ad10d5fab286/scripts/nlp_language_modeling/convert_nemo_mixtral_to_hf.py#L124), while Flake8 catches them. I strongly suggest to start using Flake8 complementary to CodeQL in CI.

**Comment by janekl** [comment]
Another example for are the unused imports `from megatron.core.utils import init_method_normal, scaled_init_method_normal` [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L94). This file is frequently edited btw. Possibly CodeQL misses this due to import guards, but I'm not sure. cc @ericharper

**Pull request merged**